### PR TITLE
Fix renderText bug for h3 elements

### LINKED ISSUE #245: renderText not working in headers
**Expected behavior** 

- changing input will change the result in header

**Actual behavior**

- following code worked in previous versions of `R` and `flexdashboard`, but in current stopped working 
- not working in 
  - `R version 3.6.1 (2019-07-05) -- "Action of the Toes"`
  - `Flexdashboard 0.5.1.1`
  - `shiny 1.4.0`

**Example**

````
--
title: "test"
output: flexdashboard::flex_dashboard
runtime: shiny
---

```{r}
library(flexdashboard)
library(shiny)
```

Inputs {.sidebar}
-------------------------------------

```{r}
selectInput(
  "month", 
  label = "Pick a Month",
  choices = month.abb, 
  selected = month.abb[2]
)

getmonth <- reactive({
  sprintf('Box 2 (%s)', input$month)
})

renderText({getmonth()})
```  

Column  
-------------------------------------

### `r renderText(getmonth())`
````

### PR DESCRIPTION
**Pull request by schloerke**
Fixes #245

~~There was a timing issue with how shiny initialized.  Because titles were shuffled around, shiny said it looked like a `shiny-bound-output` class object, but it didn't have the proper information.~~

~~`$(document).ready()` is not needed as the JS is sourced at end of body and the Flexdashboard code is sourced at the beginning of the body.  So the `setTimeout` can be dropped from https://github.com/rstudio/flexdashboard/commit/acf013d6ef58634e9283e54e5c8a18d48e8cc03b~~

-----------------------

~~Have tried on Shiny 1.4+ and Shiny 1.3.2.  (The issue was created when jQuery was bumped to v3. See https://github.com/rstudio/flexdashboard/issues/245#issuecomment-567207986)~~

### CODE REVIEW

File: R/flex_dashboard.R [hunk 1]
```diff
@@ -354,7 +354,6 @@ flex_dashboard <- function(fig_width = 6.0,
     # dashboard init script
     dashboardInitScript <- c(
        '<script type="text/javascript">',
-       '$(document).ready(function () {',
```
Comment (jjallaire) [inline R/flex_dashboard.R]: I feel like the document needs to be loaded before we run FlexDashboard.init, but perhaps your experiments prove otherwise? In any case I think this qualifies as a  very high risk change so if there is any other way to do it that would be preferred.

Comment (schloerke) [reply R/flex_dashboard.R]: 👍

Comment (jjallaire) [reply R/flex_dashboard.R]: I'm okay w/ this change if you've tested it on a bunch of static, runtime: shiny, and runtime: shiny_prerendered dashboards (including dashboards that use a variety of popular htmlwidgets, whose sizing behavior might be affected by this change).

**Comment by schloerke** [comment]
Changed approach to use and `parent.append(el.contents())` vs `parent.html(el.html)`.  

jquery links:
* [contents](https://api.jquery.com/contents/)
* [append](https://api.jquery.com/append/)

By not serializing to strings, the shiny data objects can be kept intact.  The order of execution had changed once shiny moved up to jquery v3 in shiny v1.4.0.  

These changes are only necessary to maintain the existing shiny data objects hidden in the DOM elements.  Any occurrence of `.html(x.html())` (if appropriate), was replaced with some form of `.append(el.contents())`.

Changes:
* `page headings` (bug fix)
* `h2 -> h3` titles (bug fix)
* `tab titles` (bug fix)
* `captions` (bug fix)
* `global sidebar`
* `h3 -> li`
* `chart titles`
* `valueBox`



<details>
  <summary>Jster testing document for github.com/rstudio/shinycoreci-apps testing...</summary>
---
title: "Shiny Location Kitchen Sink"
output:
  flexdashboard::flex_dashboard:
    orientation: rows
runtime: shiny
---

```{r}
library(flexdashboard)
library(shiny)
library(dygraphs)
```

# Inputs: `r renderText({getmonth()})` {.sidebar}


```{r}
selectInput(
  "month",
  label = "Pick a Month",
  choices = month.abb,
  selected = month.abb[2]
)
getmonth <- reactive({
  input$month
})
```

<div id="sidebarContent">

Sidebar Content: `r renderText(getmonth())`

</div>

```{r}
shinyjster::shinyjster_server(input, output, session)
shinyjster::shinyjster_js("
  var jst = jster();
  jst.add(Jster.shiny.waitUntilStable);
  jst.add(function() {
    // make sure on first page
    $('a[href=\"#section-p1\"]').click();
  });
  jst.add(Jster.shiny.waitUntilStable);
  jst.add(Jster.shiny.waitUntilStable);
  jst.add(Jster.shiny.waitUntilStable);
  jst.add(Jster.shiny.waitUntilStable);
  jst.add(Jster.shiny.waitUntilStable);
  jst.add(Jster.shiny.waitUntilStable);

  function assert_equal(id, txt) {
    assert_equal_qs('#section-' + id, txt);
  }
  function assert_equal_qs(qs, txt) {
    jst.add(function() {
      var el = $(qs);
      Jster.assert.isEqual(
        el.text().trim(),
        txt,
        {query_string: qs, element: el}
      );
    })
  }

  assert_equal_qs('a[href=\"#section-p1\"]', 'Page 1: Feb');
  assert_equal_qs('a[href=\"#section-p2\"]', 'Page 2: Feb');
  assert_equal('sidebarContent', 'Sidebar Content: Feb');

  assert_equal_qs('#section-hidden1 .value-output', 'VB1: Feb');
  assert_equal_qs('#section-hidden2 .value-output', 'VB2: Feb');
  assert_equal_qs('#section-hidden3 .value-output', 'VB3: Feb');

  assert_equal_qs('#section-row2 .chart-title', 'Row 2 Title: Feb');
  assert_equal_qs('#section-row2 p', 'Row 2 Content: Feb');
  assert_equal_qs('#section-row2 .chart-notes', 'Caption: Feb');

  assert_equal_qs('#section-row3 .chart-title', 'Row 3 Title: Feb');
  assert_equal_qs('#section-row3 p', 'Row 3 Content: Feb');

  assert_equal_qs('a[href=\"#section-tab1\"]', 'Tab 1 Title: Feb');
  assert_equal_qs('#section-tab1 p', 'Tab 1 Content: Feb');

  jst.add(function() {
    $('a[href=\"#section-tab2\"]').click();
  });
  jst.add(Jster.shiny.waitUntilStable);

  assert_equal_qs('a[href=\"#section-tab2\"]', 'Tab 2 Title: Feb');
  assert_equal_qs('#section-tab2 p', 'Tab 2 Content: Feb');

  // test second page
  jst.add(function() {
    $('a[href=\"#section-p2\"]').click();
  });
  jst.add(Jster.shiny.waitUntilStable);

  assert_equal_qs('#section-p2r1 .chart-title', 'P2R1 Title: Feb');
  assert_equal_qs('#section-p2r1 p', 'P2R1 Content: Feb');

  assert_equal_qs('#section-p2r2 .chart-title', 'P2R2 Title: Feb');
  assert_equal_qs('#section-p2r2 p', 'P2R2 Content: Feb');

  assert_equal_qs('a[href=\"#section-p2t1\"]', 'P2T1 Title: Feb');
  assert_equal_qs('#section-p2t1 p', 'P2T1 Content: Feb');

  jst.add(function() {
    $('a[href=\"#section-p2t2\"]').click();
  });
  jst.add(Jster.shiny.waitUntilStable);
  assert_equal_qs('a[href=\"#section-p2t2\"]', 'P2T2 Title: Feb');
  assert_equal_qs('#section-p2t2 p', 'P2T2 Content: Feb');



  jst.test();
  console.log('done!')
")
```


# Page 1: `r renderText({getmonth()})` {id=p1}


Row
-------------------------------------

### Hidden title 1 {id=hidden1}

```{r}
renderValueBox({
  valueBox({paste0("VB1: ", getmonth())}, "Value Box 1")
})
```

### Hidden title 2 {id=hidden2}

```{r}
renderValueBox({
  valueBox({paste0("VB2: ", getmonth())}, "Value Box 2")
})
```

### Hidden title 3 {id=hidden3}

```{r}
renderValueBox({
  valueBox({paste0("VB3: ", getmonth())}, "Value Box 3")
})
```

Row
-------------------------------------

### Row 2 Title:  `r renderText(getmonth())` {id=row2}

Row 2 Content: `r renderText(getmonth())`

```{r}
dygraph(ldeaths)
```

> Caption: `r renderText(getmonth())`


Row
-------------------------------------

### Row 3 Title:  `r renderText(getmonth())` {id=row3}

Row 3 Content: `r renderText(getmonth())`

```{r, fig.width=10, fig.height=7}
plot(cars)
```


Row {.tabset}
-------------------------------------

### Tab 1 Title:  `r renderText(getmonth())` {id=tab1}

Tab 1 Content: `r renderText(getmonth())`

```{r}
knitr::kable(mtcars)
```

### Tab 2 Title:  `r renderText(getmonth())` {id=tab2}

Tab 2 Content: `r renderText(getmonth())`

```{r}
renderTable({
  head(mtcars, n = 12)
})
```


# Page 2: `r renderText({getmonth()})` {id=p2}


Row
-------------------------------------

### P2R1 Title:  `r renderText(getmonth())` {id=p2r1}

P2R1 Content: `r renderText(getmonth())`

Row
-------------------------------------

### P2R2 Title:  `r renderText(getmonth())` {id=p2r2}

P2R2 Content: `r renderText(getmonth())`


Row {.tabset}
-------------------------------------

### P2T1 Title:  `r renderText(getmonth())` {id=p2t1}

P2T1 Content: `r renderText(getmonth())`

### P2T2 Title:  `r renderText(getmonth())` {id=p2t2}

P2T2 Content: `r renderText(getmonth())`

</summary>

File: inst/rmarkdown/templates/flex_dashboard/resources/flexdashboard.js [hunk 2]
```diff
@@ -402,14 +402,14 @@ var FlexDashboard = (function () {
...
-      h3.html(h2.html());
+      h3.append(h2.contents());
       h3.insertBefore(h2);
-      h2.remove();
+      h2.detatch();
 
       // wipeout h2 elements then enclose them in a single h2
       var level2 = page.find('div.section.level2');
       level2.each(function() {
-        level2.children('h2').remove();
```
Comment (wch) [inline inst/rmarkdown/templates/flex_dashboard/resources/flexdashboard.js]: Might not need to change this one to `remove()`.

File: inst/rmarkdown/templates/flex_dashboard/resources/flexdashboard.js [hunk 3]
```diff
@@ -546,8 +546,8 @@ var FlexDashboard = (function () {
       // extract the title from the h3
       var li = $('<li></li>');
       var h3 = frame.children('h3');
-      li.html(h3.html());
-      h3.remove();
```
Comment (wch) [inline inst/rmarkdown/templates/flex_dashboard/resources/flexdashboard.js]: The switch to `remove()` might not be necessary since we're doing `append()` before, not after.

File: inst/rmarkdown/templates/flex_dashboard/resources/flexdashboard.js [hunk 4]
```diff
@@ -1125,8 +1126,8 @@ var FlexDashboard = (function () {
     var h3 = container.children('h3').first();
```
Comment (wch) [inline inst/rmarkdown/templates/flex_dashboard/resources/flexdashboard.js]: Comment for this function should say "its inner contents".

File: inst/rmarkdown/templates/flex_dashboard/resources/flexdashboard.js [hunk 5]
```diff
@@ -1666,12 +1667,12 @@ window.FlexDashboardComponents.push({
       chartValue = chartValue.replace("[1] ", "");
       valueOutputSpan = valueBox.find('span.value-output').detach();
       valueBox.children().remove();
-      value.text(chartValue);
+      value.append(chartValue);
```
Comment (wch) [inline inst/rmarkdown/templates/flex_dashboard/resources/flexdashboard.js]: This should be `text()`, not `append()`.
